x/mongo/driver/operation_test.go

```diff
@@ -282,6 +282,72 @@ func TestOperation(t *testing.T) {
 			}
 		})
 	})
+	t.Run("calculateMaxTimeMS", func(t *testing.T) {
+		timeout := 5 * time.Second
+		maxTime := 2 * time.Second
+		negMaxTime := -2 * time.Second
+		shortRTT := 50 * time.Millisecond
+		longRTT := 10 * time.Second
+		timeoutCtx, cancel := internal.MakeTimeoutContext(context.Background(), timeout)
+		defer cancel()
+
+		testCases := []struct {
+			name  string
+			op    Operation
+			ctx   context.Context
+			rtt90 time.Duration
+			want  uint64
+			err   error
+		}{
+			{
+				name:  "uses context deadline and rtt90 with timeout",
+				op:    Operation{MaxTime: &maxTime},
+				ctx:   timeoutCtx,
+				rtt90: shortRTT,
+				want:  5000,
+				err:   nil,
+			},
+			{
+				name:  "uses MaxTime without timeout",
+				op:    Operation{MaxTime: &maxTime},
+				ctx:   context.Background(),
+				rtt90: longRTT,
+				want:  2000,
+				err:   nil,
+			},
+			{
+				name:  "errors when remaining timeout is less than rtt90",
+				op:    Operation{MaxTime: &maxTime},
+				ctx:   timeoutCtx,
+				rtt90: timeout,
+				want:  0,
+				err:   ErrDeadlineWouldBeExceeded,
+			},
+			{
+				name:  "errors when MaxTime is negative",
+				op:    Operation{MaxTime: &negMaxTime},
+				ctx:   context.Background(),
+				rtt90: longRTT,
+				want:  0,
+				err:   ErrNegativeMaxTime,
+			},
+		}
+		for _, tc := range testCases {
+			t.Run(tc.name, func(t *testing.T) {
```

Comment:
Done. Remind me, why is `tc := tc` necessary?

Reply:
If we don't capture the range variable, `tc` in our subtest closure will be a pointer into the array of test cases, not a single test case. That pointer is updated on every loop iteration, so by the time the subtest closure goroutine runs, the loop will (probably) have finished and `tc` will point to the last test case in the test cases array. The result is that the subtest closure wil run the _last_ test case every time, not the expected test case. That doesn't happen if the subtests run synchronously because the subtest closure will always complete before the next loop iteration.

Check out a great explanation [here](https://eli.thegreenplace.net/2019/go-internals-capturing-loop-variables-in-closures/).

Reply:
Ah very cool explanation thank you!